saw-script/src/SAWScript/Include.hs

```diff
@@ -0,0 +1,176 @@
+{-# LANGUAGE OverloadedStrings #-}
+{- |
+Module      : SAWScript.Include
+Description : Processing SAWScript include directives
+License     : BSD3
+Maintainer  : saw@galois.com
+Stability   : provisional
+-}
+
+module SAWScript.Include (Processor, processExpr, processStmts) where
+
+import qualified Data.Text as Text
+import Data.Text (Text)
+import qualified Data.IORef as IORef
+import Data.IORef (IORef)
+
+import SAWCentral.AST
+
+-- Type shorthand for the file-reader function we want.
+--
+-- (The file-reader function is passed down to avoid a dependency
+-- cycle.)
+--
+type Reader = FilePath -> IO (Either [Text] [Stmt])
+
+-- Type shorthand for the type of process*.
+type Processor a = Reader -> a -> IO (Either [Text] a)
+
+-- | Context for the pass: the reader for include files
+--   plus the error messages accumulated so far.
+--
+--   The error list is a list of lists of Text, the idea being that we
+--   cons each batch of errors we get back from the reader on the
+--   front of the outer list, and read it off at the end with @concat
+--   $ reverse errs@.
+--
+--   Note that we do not _generate_ errors, only pass on errors
+--   returned by the reader. For the moment, that does not include
+--   warnings, so there's no need to handle warnings in here either.
+--
+data Ctx = Ctx {
+    ctxReader :: Reader,
+    ctxErrs :: IORef [[Text]]
+}
+
+ctxAddErrs :: Ctx -> [Text] -> IO ()
+ctxAddErrs ctx errs = do
+    let errsRef = ctxErrs ctx
+    errlists <- IORef.readIORef errsRef
+    IORef.writeIORef errsRef (errs : errlists)
+
+
+------------------------------------------------------------
+-- Main recursive pass
+
+-- | Resolve includes, for expressions.
```

Comment:
This feels like a large amount of boilerplate; is there any way to express this using a generic traversal-type thing?